File: kirun-js/src/engine/function/system/string/TrimTo.ts
import { Schema } from '../../../json/schema/Schema';
import { Event } from '../../../model/Event';
import { EventResult } from '../../../model/EventResult';
import { FunctionOutput } from '../../../model/FunctionOutput';
import { FunctionSignature } from '../../../model/FunctionSignature';
import { Parameter } from '../../../model/Parameter';
import { Namespaces } from '../../../namespaces/Namespaces';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';
import { AbstractFunction } from '../../AbstractFunction';

export class TrimTo extends AbstractFunction {
    public static readonly PARAMETER_STRING_NAME: string = 'string';

    public static readonly PARAMETER_LENGTH_NAME: string = 'length';

    public static readonly EVENT_RESULT_NAME: string = 'result';

    protected static readonly PARAMETER_STRING: Parameter = new Parameter(
        TrimTo.PARAMETER_STRING_NAME,
        Schema.ofString(TrimTo.PARAMETER_STRING_NAME),
    );

    protected static readonly PARAMETER_LENGTH: Parameter = new Parameter(
        TrimTo.PARAMETER_LENGTH_NAME,
        Schema.ofInteger(TrimTo.PARAMETER_LENGTH_NAME),
    );

    protected static readonly EVENT_STRING: Event = new Event(
        Event.OUTPUT,
        new Map([[TrimTo.EVENT_RESULT_NAME, Schema.ofString(TrimTo.EVENT_RESULT_NAME)]]),
    );

    private signature: FunctionSignature = new FunctionSignature('TrimTo')
        .setNamespace(Namespaces.STRING)
        .setParameters(
            new Map([
                [TrimTo.PARAMETER_STRING.getParameterName(), TrimTo.PARAMETER_STRING],
                [TrimTo.PARAMETER_LENGTH.getParameterName(), TrimTo.PARAMETER_LENGTH],
            ]),
        )
        .setEvents(new Map([[TrimTo.EVENT_STRING.getName(), TrimTo.EVENT_STRING]]));

    private readonly signature = this.signature;
    public getSignature(): FunctionSignature {
        return this.signature;
    }

    public constructor() {
        super();
    }

    protected async internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {
        let inputString: string = context.getArguments()?.get(TrimTo.PARAMETER_STRING_NAME);
        let length: number = context.getArguments()?.get(TrimTo.PARAMETER_LENGTH_NAME);

        return new FunctionOutput([
            EventResult.outputOf(
                new Map([[TrimTo.EVENT_RESULT_NAME, inputString.substring(0, length)]]),
            ),
        ]);
    }
}

File: kirun-py/src/kirun_py/function/system/string/trim_to.py
from __future__ import annotations

from typing import TYPE_CHECKING

from kirun_py.function.abstract_function import AbstractFunction
from kirun_py.json.schema.schema import Schema
from kirun_py.model.event import Event
from kirun_py.model.event_result import EventResult
from kirun_py.model.function_output import FunctionOutput
from kirun_py.model.function_signature import FunctionSignature
from kirun_py.model.parameter import Parameter
from kirun_py.namespaces.namespaces import Namespaces

if TYPE_CHECKING:
    from kirun_py.runtime.function_execution_parameters import FunctionExecutionParameters


class TrimTo(AbstractFunction):

    PARAMETER_STRING_NAME: str = 'string'
    PARAMETER_LENGTH_NAME: str = 'length'
    EVENT_RESULT_NAME: str = 'result'

    def __init__(self) -> None:
        super().__init__()

        param_string = Parameter(
            self.PARAMETER_STRING_NAME,
            Schema.of_string(self.PARAMETER_STRING_NAME),
        )
        param_length = Parameter(
            self.PARAMETER_LENGTH_NAME,
            Schema.of_integer(self.PARAMETER_LENGTH_NAME),
        )
        event_string = Event(
            Event.OUTPUT,
            {self.EVENT_RESULT_NAME: Schema.of_string(self.EVENT_RESULT_NAME)},
        )

        self._signature = (
            FunctionSignature('TrimTo')
            .set_namespace(Namespaces.STRING)
            .set_parameters({
                param_string.get_parameter_name(): param_string,
                param_length.get_parameter_name(): param_length,
            })
            .set_events({event_string.get_name(): event_string})
        )

    def get_signature(self) -> FunctionSignature:
        return self._signature

    async def internal_execute(
        self, context: FunctionExecutionParameters
    ) -> FunctionOutput:
        input_string: str = context.get_arguments().get(self.PARAMETER_STRING_NAME)
        length: int = context.get_arguments().get(self.PARAMETER_LENGTH_NAME)

        return FunctionOutput([
            EventResult.output_of({self.EVENT_RESULT_NAME: input_string[:length]})
        ])
